src/github.com/getlantern/http-proxy/server.go

```diff
@@ -164,12 +183,15 @@ func (s *Server) doServe(listener net.Listener, chListenOn *chan string) error {
 
 	s.httpServer = http.Server{Handler: proxy,
 		ConnState: func(c net.Conn, state http.ConnState) {
-			if state == http.StateActive {
-				select {
-				case q <- c:
-				default:
-					fmt.Print("Oops! the connection queue is full!\n")
-				}
+			switch state {
+			case http.StateActive:
+				cb.Put(c)
+			case http.StateClosed:
+				// When go server encounters abnormal request, it
+				// will transit to StateClosed directly without
+				// the connection being withdrawed.
+				// Purge it in such case.
+				cb.Purge(c.RemoteAddr().String())
 			}
 
 			if atomic.LoadUint64(&s.numConns) >= s.maxConns {
```

Comment:
BTW, I found the code executes during each state transition. If we only want to limit connections, we can run it when `StateNew`.